packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.m

```diff
@@ -25,33 +25,29 @@ @interface FLTImagePickerPlugin () <UINavigationControllerDelegate, UIImagePicke
 @implementation FLTImagePickerPlugin {
   NSDictionary *_arguments;
   UIImagePickerController *_imagePickerController;
-  UIViewController *_viewController;
   UIImagePickerControllerCameraDevice _device;
 }
 
 + (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar {
   FlutterMethodChannel *channel =
       [FlutterMethodChannel methodChannelWithName:@"plugins.flutter.io/image_picker"
                                   binaryMessenger:[registrar messenger]];
-  UIViewController *viewController =
-      [UIApplication sharedApplication].delegate.window.rootViewController;
-  FLTImagePickerPlugin *instance =
-      [[FLTImagePickerPlugin alloc] initWithViewController:viewController];
+  FLTImagePickerPlugin *instance = [FLTImagePickerPlugin new];
   [registrar addMethodCallDelegate:instance channel:channel];
 }
 
-- (instancetype)initWithViewController:(UIViewController *)viewController {
-  self = [super init];
-  if (self) {
-    _viewController = viewController;
-  }
-  return self;
-}
-
 - (UIImagePickerController *)getImagePickerController {
   return _imagePickerController;
 }
 
+- (UIViewController *)viewController {
+  UIViewController *topController = [UIApplication sharedApplication].keyWindow.rootViewController;
```

Comment:
this is great! By the way, while we are here, can we also make it work for navigation controller? (pushed viewcontrollers)

Reply:
this solution will work in this case as well.
According to the [documentation](https://developer.apple.com/documentation/uikit/uiimagepickercontroller) you shouldn't push UIImagePickerController, but present it modally instead.

So in the situation that FlutterViewController is pushed into an existing navigation controller, e.g.:

```
Window
  rootViewController
    someOtherViewController
      navigation controller (with some native view controller and FlutterVC)
         image picker controller 

not

Window
  rootViewController
    someOtherViewController
      navigation controller (with some native view controller and FlutterVC, image picker controller)
          
```

I was worried about iPad which for a long time required popover presentation for albums, but since iOS 13 it presents modally as well without an issue.

Reply:
@cyanglaz the solution I made will in all cases you are concerned about.

UINavigationController, UITabBarController they both manage internally some view controllers, **but** they themselves inherit from UIViewControllerr and either of them can be window's root view controller, or presented on some existing View Controller of any type.

The topViewController cannot be pushed in the navigation stack. In that case, UINavigationCoontroller is topViewController and some other push view controller is managed by UINavigationController. And with that arrangement, my solution will correctly display the Image Picker on top of the navigation controller.

**Trust me** - in my company we have already very complex navigation - soome legacy Drawer Controller (in iOS), multiple Navigation controllers, TabBarControllers and Flutter V.C. is displayed on top of all of this. And my solution has been proven successful.

Reply:
according to https://developer.apple.com/documentation/uikit/uiviewcontroller/1621407-presentedviewcontroller?language=objc the `presentedViewController` is only the top view controller when it is presented with `presentViewController:animated:completion:`. 
What if the top view controller is presented by [pushViewController](https://developer.apple.com/documentation/uikit/uinavigationcontroller/1621887-pushviewcontroller?language=objc)? Does this still work?

Reply:
@cyanglaz I am already repeating myself, but let me rephrase.

`presentViewController` is a method of `UIViewController`
UINavigationController, UITabBarController, MySweetViewController - they are all subclasses of UIViewController, therefore they can be presented on any other view controller, or some other view controller can be presented on top of them.

The top view controller **presented** by push view controller - such a thing doesn't exist.
You can have UINavigationController either as a root of the window, or presented on some other view controller. And Navigation Controller manages the stack of view controllers. These view controllers shouldn't present any other view controller (because they are managed by NC). Instead, you either push another VC in NC (horizontal), or present VC on top of NC (vertical).

Moreover - if UINavigationController is one of the children of UITabBarController, you shouldn't present any VC on top of that NC, you should present on top of UITabBarController instead.

Reply:
Ok. I'm interested in this case you presented 
```
example2:
window
  navigation controller (home, details)
     image picker
```

In this case, when you go through the code, what is the top view controller? Is it the `navigation controller`? 
And the image picker is going to use the navigation controller as the "viewController", regardless which view controller is the on the top of the stack of the navigation controller? Am I understanding correctly?

Reply:
exactly, image picker will be correctly presented on navigation controller. The contents of navigation controller doesn't matter at all.